src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -510,14 +562,180 @@ internal void ReadToken(string encodedJson)
+
+                ReadOnlySpan<char> headerSpan = encodedTokenSpan.Slice(0, Dot1);
+                if (headerSpan.IsEmpty)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogMessages.IDX14307));
+
+                // right number of dots for JWE (4)
+                byte[] headerAsciiBytes = new byte[headerSpan.Length];
+                Encoding.ASCII.GetBytes(headerSpan, headerAsciiBytes);
+
+                HeaderAsciiBytes = headerAsciiBytes;
+
+                try
+                {
+                    Header = CreateHeaderClaimSet(Base64UrlEncoder.UnsafeDecode(headerSpan).AsSpan());
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(
+                        LogMessages.IDX14102,
+                        LogHelper.MarkAsUnsafeSecurityArtifact(headerSpan.ToString(), t => t.ToString())),
+                        ex));
+                }
+
+                // delegating retrieving encrypted Key to the getter on EncryptedKey
+                ReadOnlySpan<char> encryptedKeyBytes = encodedTokenSpan.Slice(Dot1 + 1, Dot2 - Dot1 - 1);
+                if (!encryptedKeyBytes.IsEmpty)
+                    EncryptedKeyBytes = Base64UrlEncoder.UnsafeDecode(encryptedKeyBytes);
+
+                ReadOnlySpan<char> initializationVectorSpan = encodedTokenSpan.Slice(Dot2 + 1, Dot3 - Dot2 - 1);
+                if (initializationVectorSpan.IsEmpty)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogMessages.IDX14308));
+
+                try
+                {
+                    InitializationVectorBytes = Base64UrlEncoder.UnsafeDecode(initializationVectorSpan);
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogMessages.IDX14309, ex));
+                }
+
+                ReadOnlySpan<char> authTagSpan = encodedTokenSpan.Slice(Dot4 + 1);
+                if (authTagSpan.IsEmpty)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogMessages.IDX14310));
+
+                try
+                {
+                    AuthenticationTagBytes = Base64UrlEncoder.UnsafeDecode(authTagSpan);
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogMessages.IDX14311, ex));
+                }
+
+                ReadOnlySpan<char> cipherTextSpan = encodedTokenSpan.Slice(Dot3 + 1, Dot4 - Dot3 - 1);
+                if (cipherTextSpan.IsEmpty)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogMessages.IDX14306));
+
+                try
+                {
+                    CipherTextBytes = Base64UrlEncoder.UnsafeDecode(cipherTextSpan);
+                }
+                catch (Exception ex)
+                {
```

Comment:
>{ [](http://example.com/codeflow?start=16&length=1)

wonder if this can be broken down into smaller methods to test consolidated parts.

Reply:
ReadToken is a large method. Would be good to introduce smaller methods for readability and testing. However, it would additional calls to the call graph.

Reply:
Keeping the method as it in.